crates/bevy_egui/src/systems.rs

```diff
@@ -0,0 +1,339 @@
+        primary_window_id,
+        
+        current_tick,
+        previous_input
+    } = &mut *state;
+
+    // Uses the pressed state of the latest left mouse event, otherwise the previous state
+    let mouse_down = mouse_button_reader
+        .find_latest(&mouse_button_events, |input| {
+            input.button == MouseButton::Left
+        })
+        .map_or(previous_input.mouse_down, |input| {
+            input.state.is_pressed()
+        });
+
+    let scroll_delta = mouse_wheel_reader.iter(&mouse_wheel_events)
+        .fold(egui::vec2(0.0, 0.0), |delta, MouseWheel { x, y, .. }| {
+            delta + egui::vec2(*x, *y)
+        });
+
+    let screen_size = window_resized_reader.find_latest(&window_resized_events, |resized| {
+            resized.id == *primary_window_id
+        })
+        .map_or(previous_input.screen_size, |WindowResized { width, height, .. }| {
+            egui::vec2(*width as _, *height as _)
+        });
+
+    let mouse_pos = cursor_moved_reader
+        .find_latest(&cursor_moved_events, |cursor| {
+            cursor.id == *primary_window_id
+        })
+        .map_or(previous_input.mouse_pos, |CursorMoved { position, .. }| {
+            // Bevy has origin as bottom left, egui expects origin to be top left, so we do the math to change it
+            Some(egui::pos2(position.x(), screen_size.y - position.y()))
+        });
+
+    let events = keyboard_input_reader.iter(&keyboard_input_events)
+        .fold(Vec::<egui::Event>::new(), |mut events: Vec<egui::Event>, KeyboardInput { key_code, state, .. }| {
+            if let Some(key_code) = key_code {
+                // Handle special cased key combos like cut/copy/paste
+                if let Some(event) = match key_code {
+                    KeyCode::Copy => Some(Event::Copy),
+                    KeyCode::Cut => Some(Event::Cut),
+                    KeyCode::Paste => {
+                        tracing::warn!("Paste event isn't implemented in bevy_egui! We need to figure out how to get stuff from the clipboard...");
+                        None
+                    }
+                    _ => None
+                } {
+                    events.push(event);
+                } else if let Some(key) = key_code.into_egui_key() {
+                    events.push(Event::Key {
+                        key,
+                        pressed: state.is_pressed()
+                    });
+                } else {
+                    tracing::info!("Key presses are currently not really given to egui! Pressed: {:?}", key_code);
+                }
+            }
+
+            events
+        });
+
+    let new_input = egui::RawInput {
+        mouse_down,
+        mouse_pos,
+        scroll_delta,
+        screen_size,
+        events,
+        ..*previous_input
+    };
+
+    *previous_input = new_input.clone();
+    egui_input.raw_input = new_input;
+}
+
+trait IntoEguiKey {
+    fn into_egui_key(self) -> Option<egui::Key>;
+}
```

Comment:
Do you really need a new trait? Why can't a private function work just fine? No need to use and implement an extension trait where you don't need to. Rust has private functions! :)

Reply:
While I understand your point... I personally tend to prefer this style myself. It's nicer calling it from the receiver I think!